scipy/spatial/transform/rotation.py

```diff
@@ -1,12 +1,55 @@
 from __future__ import division, print_function, absolute_import
 
+__all__ = ['Rotation']
+
 import numpy as np
+import scipy.linalg
 
 class Rotation(object):
     """Rotation in 3 dimensions.
 
     This class will include initializers from different representations, converters
     and some useful algorithms such as quaternion slerp and rotation estimation.
+
     """
-    def __init__(self):
-        pass
+    def __init__(self, quaternions):
+        """Initialize class from possible normalized quaternions.
+
+        Parameters
+        ----------
+        quaternions : (N, 4) numpy.array
+                      Each row is a quaternion stored in scalar-last
+                      (x, y, z, w) format.
+
+        """
+        # Each row should have 4 numbers
+        assert(quaternions.shape[1] == 4)
+
+        # L2 norm of each row
+        norms = scipy.linalg.norm(quaternions, axis=1)
+
+        # Divide each row by its norm.
+        # In case quat is [4 x 4], ensure norm is broadcasted along each column
+        normalised_quat = quaternions / norms[:, None]
+        self._quat = normalised_quat
+
+    @classmethod
+    def from_quaternion(cls, quat):
+        """Initialize Rotation from possibly unnormalized quaternions
+
+        This classmethod normalizes the input quaternions and returns a
+        `Rotation` object.
+
+        Parameters
+        ----------
+        quat : (N, 4) or (4,) numpy.array
+               Each row is a (possibly non-unit norm) quaternion in scalar-last
+               (x, y, z, w) format.
+
+        """
+
+        # If a single quaternion is given, convert it to a 2D 1 x 4 matrix
+        if quat.shape == (4,):
```

Comment:
Here are my thought about storage. So you decided to store quaternions in stack always. What this leads to in my opinion:

1. You need to store a flag internally like `self._single` and if `self._single == True` then return a single object in `as_...`.
2. This part I'm not 100% sure. I would guess that applying vectorized approach for very few objects (especially 1) is not a great idea in terms of performance. On the other hand this is not measure. So what would be your strategy for this: 
   a. Code only vectorized approach.
   b. Code both approaches right away. And then either 1) iterate over objects and apply a scalar approach or 2) apply vectorized approach, all depending on performance benchmarking for different number of rotation objects (like dcm, quat, etc. )
   c. Code one of the approaches first and came back later to implement another if you have time.

I might guess that you should first experiment with it, say, for `from_dcm` code. See the performance difference depending on the number of rotation objects and then choose one of the approaches.

What do you think?

Reply:
1. Can you tell me the purpose of the `_single` flag? In my mind, we can just check the shape of the `_quat` matrix to see how many quaternions are there.

Regarding your second point: (I'm assuming that by _stack_ you are referring to the `_quat` variable and by _objects_ on the stack you are referring to each row of `_quat`? ).
If I understand your thought process correctly, the concern is that vectorized functions like `scipy.linalg.norm` will have a significant performance overhead when dealing with very few rotations. For this, I propose a `threshold` number of quaternions in a single `Rotation` object. If the number of rotations is more than this threshold, then we apply vectorized operations, otherwise we iterate over all rows in `_quat` and do our operations. This threshold will have to be chosen according to the benchmarks and various operations in our implementation.

What I'm thinking is that I'll implement the vectorized approach for now. Once this function is implemented, I'll perform some benchmarking tests (mainly checking the speed of the `scipy.linalg.norm` function) to see if there are performance impacts, and at what number can we safely use vectorized approaches. Is that ok?

Reply:
> Can you tell me the purpose of the _single flag? In my mind, we can just check the shape of the _quat matrix to see how many quaternions are there.

I added the explanation later, perhaps you didn't see it. So the problematic case is `(1, 4)` input )

> If I understand your thought process correctly, the concern is that vectorized functions like scipy.linalg.norm will have a significant performance overhead when dealing with very few rotations. For this, I propose a threshold number of quaternions in a single Rotation object. If the number of rotations is more than this threshold, then we apply vectorized operations, otherwise we iterate over all rows in _quat and do our operations. This threshold will have to be chosen according to the benchmarks and various operations in our implementation.

Your general understanding is correct. However I'm speaking about other cases. Well, it isn't strongly related to `_quat`, my bad here. 

For `from_quat` I don't see any problem in vectorized approach. But consider `from_dcm` --- as you may recall there are 4 different cases. If you do the vectorized approach you will write something like `mask_1 = (boolean mask for case 1)`, `mask_2 = ...`, `mask_3 = ...`, `mask_4 = ...`. Then you write something like `quat_[mask_1] = ...`.  You can see that this is quite strange and possibly inefficient approach when we have only few dcms to convert. The same issues will arise in other conversion routines.

From our discussion I suggest to solve this case by case. In difficult cases like `from_dcm` I believe both approaches might be necessary. So I suggest to experiment with it in `from_dcm` to see what is the difference in performance.

Is it clear what I said?

Reply:
Oh yes, that is much clearer, thanks. I was aware of the 4 cases, I just thought this concern was related to `_quat`. I'll keep that in mind when I'm writing the `from_dcm` function.